Lib/test/test_exception_group.py

```diff
@@ -0,0 +1,746 @@
+
+
+class BadConstructorArgs(unittest.TestCase):
+    def test_bad_EG_construction__too_many_args(self):
+        MSG = 'function takes exactly 2 arguments'
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup('no errors')
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup([ValueError('no msg')])
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup('eg', [ValueError('too')], [TypeError('many')])
+
+    def test_bad_EG_construction__bad_message(self):
+        MSG = 'argument 1 must be str, not '
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup(ValueError(12), SyntaxError('bad syntax'))
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup(None, [ValueError(12)])
+
+    def test_bad_EG_construction__bad_excs_sequence(self):
+        MSG = 'argument 2 must be a non-empty sequence'
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup('errors not sequence', {ValueError(42)})
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup("eg", [])
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup("eg", None)
+
+    def test_bad_EG_construction__nested_non_exceptions(self):
+        MSG = 'Item [0-9]+ of argument 2 is not an exception instance'
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup('expect instance, not type', [ValueError]);
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup('bad error', ["not an exception"])
+        with self.assertRaisesRegex(TypeError, MSG):
+            _ = ExceptionGroup('eg1',
+                               [ExceptionGroup('eg2', ["not an exception"])])
+
+
+class InstanceCreation(unittest.TestCase):
+    def test_EG_wraps_Exceptions__creates_EG(self):
+        excs = [ValueError(1), TypeError(2)]
+        self.assertIs(type(ExceptionGroup("eg", excs)), ExceptionGroup)
+
+    def test_BEG_wraps_Exceptions__creates_EG(self):
+        excs = [ValueError(1), TypeError(2)]
+        self.assertIs(type(BaseExceptionGroup("beg", excs)), ExceptionGroup)
+
+    def test_EG_wraps_BaseException__raises_TypeError(self):
+        MSG= "Cannot nest BaseExceptions in an ExceptionGroup"
+        with self.assertRaisesRegex(TypeError, MSG):
+            eg = ExceptionGroup("eg", [ValueError(1), KeyboardInterrupt(2)])
+
+    def test_BEG_wraps_BaseException__creates_BEG(self):
+        beg = BaseExceptionGroup("beg", [ValueError(1), KeyboardInterrupt(2)])
+        self.assertIs(type(beg), BaseExceptionGroup)
+
+    def test_EG_subclass_wraps_anything(self):
+        class MyEG(ExceptionGroup):
+            pass
+        self.assertIs(
+            type(MyEG("eg", [ValueError(12), TypeError(42)])), MyEG)
+        self.assertIs(
+            type(MyEG("eg", [ValueError(12), KeyboardInterrupt(42)])), MyEG)
```

Comment:
For user-defined subclasses, we have to punt on the special behavior. And yes, user classes should probably do what you show, but we don't want to enforce it. (That would be making assumptions.)

Reply:
Fair enough!

IMO defining custom `ExceptionGroup` subclasses should be *very rare* anyway (maybe just Hypothesis and Trio?), and we can certainly deal with whatever complexity that requires.

Reply:
Eh, why would Hypothesis need to subclass EG? While we were called on arbitrarily *forbidding* subclassing (as we did in an earlier PEP draft) I don't see why it would be useful. What is your use case? Maybe there's something we can add to the design to avoid the need? (For Trio I could imagine they'd want to provide some kind of backwards compatibility with Trio's current solution.)

Reply:
Maybe a subclass to annotate the nested exceptions?   See https://discuss.python.org/t/accepting-pep-654-exception-groups-and-except/10813/9

Reply:
Exactly! Maybe we *don't* need to, but my motivation is that Hypothesis wants to track and report some additional information (such as the minimal reproducing example) for each exception.  For example:

```python
from hypothesis import given, strategies as st

@given(st.integers())
def test_multiple_failures(x):
    assert x < 100, "This number is too big"
    assert x > -100, "This number is too small"
```

```python-traceback
Traceback (most recent call last):
  ...
hypothesis.errors.MultipleFailures: Hypothesis found 2 distinct failures.
----------------------------------------------------- Hypothesis -----------------------------------------------------
Falsifying example: test_multiple_failures(x=-100)
Traceback (most recent call last):
  File "example.py", line 7, in test_multiple_failures
    assert x > -100, "This number is too small"
AssertionError: This number is too small

Falsifying example: test_multiple_failures(x=100)
Traceback (most recent call last):
  File "example.py", line 6, in test_multiple_failures
    assert x < 100, "This number is too big"
AssertionError: This number is too big
```

Currently `Falsifying example: test_multiple_failures(x=100)` is one [fragment to report](https://github.com/HypothesisWorks/hypothesis/blob/710699c3cdbfcb3a342de08f426c1c8906c66cba/hypothesis-python/src/hypothesis/core.py#L1130-L1143), i.e. print-if-not-in-quiet-mode, but we could turn them into a single string easily enough.

Reply:
Okay, I'll bite. What would you like to see supported upstream? A separate metadata item for each sub-exception? I'd say if you don't want to store that on the sub-exception in some dunder-ish attribute, you could indeed subclass EG to add that. There's an API so you can replicate this whenever an EG is being filtered. But I don't think we're going to support metadata that automatically gets replicated based on just this example -- in most cases I suspect you can just store an extra attribute on the individual exceptions.

Reply:
> I'd say if you don't want to store that on the sub-exception in some dunder-ish attribute, you could indeed subclass EG to add that.

I'm very happy to store it on sub-exceptions like this; I just anticipated that I'd need to subclass in order to get the lets-call-it-`__note__` displayed nicely along with the traceback.

Maybe I can just set that and add support in e.g. Pytest as a not-quite-a-standard convention, though.

Reply:
@Zac-HD: Why don't you play with the implementation (linked from the PEP) and see if you can figure out to get it to work the way you want, rather than us having this side discussion in Irit's PR.